keps/sig-cli/3659-kubectl-apply-prune/README.md

```diff
@@ -0,0 +1,1365 @@
+
+This list may be extended in the future as use cases are presented. Keeping it minimal has significant benefits for the performance of any commands that list the parent objects themselves. Specific tools may further limit the types they support, but any tooling designed to identify all ApplySets must consider the exact list set out above.
+
+While a purpose-made cluster-scoped CRD is a logical choice, the specification has no opinion on this, so as to accommodate the many and various choices existing tools have already made. In choosing types to support as parents, tools should consider what permissions their target users typically have on their clusters; for instance, they may not have permissions to install CRDs.
+
+#### Labels and annotations
+
+ApplySets MUST also have a “parent object” in the cluster. The ApplySet object MUST be labeled with:
+
+```yaml
+applyset.k8s.io/id: <applysetkey> # REQUIRED
+```
+
+The `applyset.k8s.io/id` label is what makes the object an ApplySet parent object. Its value MUST match the value of `applyset.k8s.io/part-of` on the member objects (see [ApplySet Member Objects](#labels)), and MUST comply with the naming constraints outlined in [ApplySet Naming](#applyset-naming).
+
+Additionally, ApplySet parents MUST be labelled with:
+
+```yaml
+applyset.k8s.io/tooling: <toolId> # REQUIRED
+```
+
+The value should be something like `kubectl/v1.27` or `helm/v3` or `kpt/v1.0.0`,
+i.e. `<toolname>/<semver>`, and tooling should refuse to mutate ApplySets belonging to other tools.
+For more background and guidance on this topic, see the [interoperability](#tooling-interoperability) section.
+
+ApplySets MAY have an annotation extending their scope, which MUST be respected by all tools if present:
+
+```yaml
+applyset.k8s.io/additional-namespaces: <ns>[,] # OPTIONAL
+```
+
+The `applyset.k8s.io/additional-namespaces` annotation extends the scope of the ApplySet. By default, the scope of an ApplySet is its parent object's namespace. When the parent is cluster-scoped but refers to namespace kinds (see below), or when the set spans multiple namespaces (which is not recommended, but allowed), this annotation can be used to extend the ApplySet's scoped to the listed namespaces. As cross-namespace ApplySets are not particularly encouraged, we do not currently optimize this further.  In particular, we do not specify the GKs per namespace.  We can add more annotations in future should the need arise.
+
+The value of this annotation is a comma-separated list of the names of the namespaces (other than the ApplySet namespace) in which objects are found, for example `default,kube-system,ns1,ns2`. We reserve the empty value. As with `applyset.k8s.io/contains-group-kinds`, this list of namespaces must be sorted alphabetically, and should be minimal (other than transiently during ApplySet mutations).
```

Comment:
It's very hard to enforce this data format in an annotation that can appear on multiple objects, maybe this is why it feels bad to me.

Reply:
@liggitt may have thoughts on this. IMO comma delimited, sorted values aren't a good fit an annotation, especially when the versioning plan is to prefix the value with "v2:", if I understood the following paragraphs correctly.

Reply:
We're definitely open to suggestions - that's what the KEP process is for :-)  This annotation is less critical than the labels.  Katrina pointed out that if we just had the labels and dumped both the "just these kinds" and the "just these namespaces" annotations, the system would still work - it would just be less efficient and require more permissions.  We do have an "fsck" command to detect (and likely repair) inconsistencies between the labels and the annotations.

Reply:
I'm still trying to get oriented with the data flow being proposed, asked a couple questions in that dimension